File: d2/d2b.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#define MAX_LINE_LENGTH 1024
#define MAX_RANGE_LENGTH 32

bool is_code_invalid(long code) {
	char codestr[15];
	char target[15];
	char chunk[15];
	sprintf(codestr, "%li", code);

	int code_len = strlen(codestr);
	
	for (int num_chars=1; num_chars<=code_len/2; num_chars++) {
		if ((code_len % num_chars) == 0) {
			// generate target code
			target[0]='\0';
			// get substring chunk to repeat
			strcpy(chunk, codestr);
			chunk[num_chars]='\0';
			// create target from repeated chunk
			for (int n=0; n<code_len/num_chars; n++) {
				strcat(target, chunk); 
			}
			// test if invalid
			if (strcmp(codestr, target) == 0) {
				return true;
			}
		}
	}
	return false;
}

int main(void) {
	char buffer[MAX_LINE_LENGTH];
	
	int invalid_count = 0;
	long invalid_sum = 0;

	long start, stop;
	
	scanf("%s", buffer);
	// char range[MAX_RANGE_LENGTH];
	char *range = strtok(buffer, ",");
	// loop over input
	while (range != NULL) {
		sscanf(range, "%li-%li", &start, &stop);
		// printf("%li-%li ", start, stop);
		for (long code = start; code <= stop; code++) {
			if (is_code_invalid(code)) {
				// printf("%li ", code);
				invalid_count++;
				invalid_sum = invalid_sum + code;
			}
		}
		// printf("\n");
		range = strtok(NULL, ",");
	}
	printf("%i invalid codes summing to %li.\n", invalid_count, invalid_sum);
}

File: d2/d2b.py
#/usr/bin/env python
import sys
import re
invalid_count=0
invalid_sum=0
search_ranges = sys.stdin.read().strip().split(',')
search_ranges = [tuple([int(val) for val in x.split('-')]) for x in search_ranges]

def is_code_invalid(code):
        code = str(code)
        for repeated_characters in range(1, len(code)//2+1):
            repeats = len(code)//repeated_characters
            if code == code[:repeated_characters]*repeats:
                return True
        return False

for start, stop in search_ranges:
    #print(f'searching range {start}-{stop}.')
    for code in range(start, stop + 1):
        #if bool(re.fullmatch(r'^(.+)\1+$', str(code))):
        if is_code_invalid(code):
            invalid_count += 1
            invalid_sum += code
            #print(f'code {code} is invalid!')
print(f'Num invalid codes: {invalid_count}')
print(f'Sum invalid codes: {invalid_sum}')
